src/core/border_router/routing_manager.cpp

```diff
@@ -0,0 +1,796 @@
+
+    OT_ASSERT(aFirstPrefix.GetLength() == aSecondPrefix.GetLength());
+
+    matchedLength =
+        Ip6::Prefix::MatchLength(aFirstPrefix.GetBytes(), aSecondPrefix.GetBytes(), aFirstPrefix.GetBytesSize());
+
+    return matchedLength < aFirstPrefix.GetLength() &&
+           aFirstPrefix.GetBytes()[matchedLength / CHAR_BIT] < aSecondPrefix.GetBytes()[matchedLength / CHAR_BIT];
+}
+
+bool RoutingManager::IsValidOmrPrefix(const Ip6::Prefix &aOmrPrefix)
+{
+    // Accept ULA prefix with length of 64 bits and GUA prefix.
+    return (aOmrPrefix.mLength == kOmrPrefixLength && aOmrPrefix.mPrefix.mFields.m8[0] == 0xfd) ||
+           (aOmrPrefix.mLength >= 3 && (aOmrPrefix.GetBytes()[0] & 0xE0) == 0x20);
+}
+
+bool RoutingManager::IsValidOnLinkPrefix(const Ip6::Prefix &aOnLinkPrefix)
+{
+    // Accept ULA prefix with length of 64 bits and GUA prefix.
+    return (aOnLinkPrefix.mLength == kOnLinkPrefixLength && aOnLinkPrefix.mPrefix.mFields.m8[0] == 0xfd) ||
+           (aOnLinkPrefix.mLength >= 3 && (aOnLinkPrefix.GetBytes()[0] & 0xE0) == 0x20);
+}
+
+void RoutingManager::HandleRouterAdvertisementTimer(Timer &aTimer)
+{
+    aTimer.GetOwner<RoutingManager>().HandleRouterAdvertisementTimer();
+}
+
+void RoutingManager::HandleRouterAdvertisementTimer(void)
+{
+    otLogInfoBr("Router Advertisement timer triggered");
+
+    EvaluateRoutingPolicy();
+}
+
+void RoutingManager::HandleRouterSolicitTimer(Timer &aTimer)
+{
+    aTimer.GetOwner<RoutingManager>().HandleRouterSolicitTimer();
+}
+
+void RoutingManager::HandleRouterSolicitTimer(void)
+{
+    otLogInfoBr("Router Solicit timeouted");
+
+    if (mRouterSolicitCount < kMaxRtrSolicitations)
+    {
+        uint32_t nextSolicitationDelay;
+        otError  error;
+
+        error = SendRouterSolicitation();
+        ++mRouterSolicitCount;
+
+        if (error == OT_ERROR_NONE)
+        {
+            otLogDebgBr("Successfully sent %uth Router Solicitation", mRouterSolicitCount);
+        }
+        else
+        {
+            otLogCritBr("failed to send %uth Router Solicitation: %s", mRouterSolicitCount,
+                        otThreadErrorToString(error));
+        }
+
+        nextSolicitationDelay =
+            (mRouterSolicitCount == kMaxRtrSolicitations) ? kMaxRtrSolicitationDelay : kRtrSolicitationInterval;
+
+        otLogDebgBr("Router Solicitation timer scheduled in %u seconds", nextSolicitationDelay);
+        mRouterSolicitTimer.Start(nextSolicitationDelay * 1000);
+    }
+    else
+    {
+        // Re-evaluate our routing policy and send Router Advertisement if necessary.
+        EvaluateRoutingPolicy();
+    }
+}
+
+void RoutingManager::HandleDiscoveredOnLinkPrefixInvalidTimer(Timer &aTimer)
+{
+    aTimer.GetOwner<RoutingManager>().HandleDiscoveredOnLinkPrefixInvalidTimer();
+}
+
+void RoutingManager::HandleDiscoveredOnLinkPrefixInvalidTimer(void)
+{
+    otLogInfoBr("invalidate discovered on-link prefix: %s", mDiscoveredOnLinkPrefix.ToString().AsCString());
```

Comment:
nit: prefer to be consistent about capitalizing (or not) the beginning of the log message